source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,816 @@
+
+      // if this is the first time we get here, set the last output time
+      // to the current time - output_interval. this makes sure we
+      // always produce data during the first time step
+      if (std::isnan(last_output_time))
+        last_output_time = this->get_time() - output_interval;
+
+      // If it's not time to generate an output file or we do not write output
+      // return early.
+      if (this->get_time() < last_output_time + output_interval && this->get_time() != end_time)
+        return std::make_pair("","");
+
+      if (output_file_number == numbers::invalid_unsigned_int)
+        output_file_number = 0;
+      else
+        ++output_file_number;
+
+      // Now prepare everything for writing the output and choose output format
+      std::string particle_file_prefix_master = this->get_output_directory() +  "particle_CPO/particles-" + Utilities::int_to_string (output_file_number, 5);
+      std::string particle_file_prefix_content_raw = this->get_output_directory() +  "particle_CPO/CPO-" + Utilities::int_to_string (output_file_number, 5);
+      std::string particle_file_prefix_content_draw_volume_weighting = this->get_output_directory() +  "particle_CPO/weighted_CPO-" + Utilities::int_to_string (output_file_number, 5);
+
+      const auto &particle_handler = this->get_particle_world().get_particle_handler();
+
+      std::stringstream string_stream_master;
+      std::stringstream string_stream_content_raw;
+      std::stringstream string_stream_content_draw_volume_weighting;
+
+      string_stream_master << "id x y" << (dim == 3 ? " z" : " ") << " olivine_deformation_type" << std::endl;
+
+      // get particle data
+      bool wrote_weighted_header = false;
+      bool wrote_unweighted_header = false;
+      for (typename dealii::Particles::ParticleHandler<dim>::particle_iterator it = particle_handler.begin(); it != particle_handler.end(); ++it)
+        {
+
+          AssertThrow(it->has_properties(),
+                      ExcMessage("No particle properties found. Make sure that the CPO particle property plugin is selected."));
+
+
+
+          unsigned int id = it->get_id();
+          const ArrayView<double> &properties = it->get_properties();
+
+          const Particle::Property::ParticlePropertyInformation &property_information = this->get_particle_world().get_property_manager().get_data_info();
+
+          AssertThrow(property_information.fieldname_exists("cpo mineral 0 type") ,
+                      ExcMessage("No CPO particle properties found. Make sure that the CPO particle property plugin is selected."));
+
+
+
+          const unsigned int cpo_data_position = property_information.n_fields() == 0
+                                                 ?
+                                                 0
+                                                 :
+                                                 property_information.get_position_by_field_name("cpo mineral 0 type");
```

Comment:
Move lines 223-234 outside the loop over all particles, they are the same for all particles.